### ISSUE TITLE
Freezes when going back and forth between activites

### ISSUE THREAD

**Issue by erf** [mapbox/mapbox-gl-native#2018]
If you add a button to the main view in the test app, and open an arbitrary activity on a click event and then closes it, moves the map and click the button again, the app freezes on my nexus 7 (2012).
I made a simple test case here: 
https://github.com/erf/mapbox-gl-native/tree/freezes

**Comment by erf** [comment]
BTW I did not get any log error events.. but this is a MAJOR issue, if creating an app that uses more than one Activity makes the app freeze up, than the library is pretty much useless. Does anyone else experience this?

**Comment by erf** [comment]
This also seem to happen if i moved the map first and then click on a button which opens the new activity. I don't see any information in the Log regarding this. I suspected this may be a memory issue when opening the map several times, so i even tried to make the mapview static, but with same result. I suspect now that some state is getting set when you move the map, but i have no idea what causes the crash.

**Comment by erf** [comment]
I just made a simple example now on latest 4025da6 and this seem to work now! I close this for now.

**Comment by erf** [comment]
This issue is back after i added loading mapView as a fragment described in #2042

**Comment by tobrunvannuland** [comment]
Was able to reproduce this occasionally with following setup:
Genymotion - Google Nexus 5.0.0 image - 0.0.1-SNAPSHOT

Actions:
Open activity with MapView, scroll-zoom actions like a maniac, open other Activity containing a MapView. Application becomes unresponsive and screens turns black. Only option is to kill-app by swiping it from recents.

Used Activities:
StyledMapActivity -  SearchMapActivity

[Full stacktrace](http://pastebin.com/vBK8Cx5S) - [code](https://github.com/tobrunvannuland/mapbox-gl-native/tree/playground/android/java/MapboxGLAndroidPlaygroundApp)

**Comment by erf** [comment]
I think the app crashes when trying to open a new activity whilst the MapView is still loading tiles

**Comment by ljbade** [comment]
I also wonder if it is related to state saving.

**Comment by ljbade** [comment]
@kkaefer Is there anything in the C++ layer to be aware of when there are multiple instances of a map?

**Comment by kkaefer** [comment]
@ljbade it should be possible and I'm not aware of an issues, but of course there could be bugs when using multiple instances.

**Comment by incanus** [comment]
We are handling multiple maps in iOS ok right now.

**Comment by ljbade** [comment]
Hmm going to have to create a simple test app to see if I can reproduce it.

@incanus Do you recall if there were any specific fixes/changes that perhaps did not make it to Android to enable multiple maps?

**Comment by tobrunvannuland** [comment]
@ljbade a simple test app can be found [here](https://github.com/tobrunvannuland/mapbox-gl-native/tree/playground/android/java/MapboxGLAndroidPlaygroundApp)

**Comment by ljbade** [comment]
Thanks @tobrunvannuland

**Comment by erf** [comment]
Simple test case:
Try to have a map in both activity A and B. Open activity B from A and click in the map before all the tiles are loaded.

**Comment by ljbade** [comment]
I done some reading at it appears the Android system does not support two `SurfaceView`s active at the same time.

So this bug might be due to limitations as a result of using `SurfaceView`.

Perhaps we will need to switch to `TextureView` to support more complex layouts.

**Comment by ljbade** [comment]
We are pushing `TextureView` to `0.2.0` so we should wait to fix this bug then?

@bleege ?

**Comment by bleege** [comment]
> Perhaps we will need to switch to TextureView to support more complex layouts.

@ljbade Thanks for looking into this.  Before we can make a decision on milestone can you provide more detail as to why you believe `TextureView` would solve this issue?  If there's a firm reason why you believe it should then we can definitely look to include #2244 in `0.1.0`.

**Comment by ljbade** [comment]
@bleege Basically we use `SurfaceView`. It is well documented the Android system does not support two `SurfaceView`s on the screen at the same time. This would include two different `Activitiy` where there might be some sort of transition or delay between the new `MapView` being created and the old `MapView` getting deleted.

`TextureView` does not have this limitation.

**Comment by ljbade** [comment]
Although we also need to test if `TextureView` will fix this bug, in case it is something related to the native layer.

**Comment by ljbade** [comment]
@erf Do you mind recreating the bug against the `TextureView` implementation in https://github.com/mapbox/mapbox-gl-native/tree/2244-texture-view branch?

I would like to know if this is a solution.

**Comment by erf** [comment]
Will do. Just really busy right now..

**Comment by tobrun** [comment]
I will try to do some testing as well

**Comment by tobrun** [comment]
I have forked the branch [here](https://github.com/tobrun/mapbox-gl-native/tree/2244-texture-view). I replaced the FAB action with 

```java
startActivity(new Intent(MainActivity.this, MainActivity.class));
```

This action will allow us to stack multiple Activities on top of each other. A basic test without scrolling/zooming doesn't cause any crash or freeze.  (Tested with stack of 10 Activities).

But I was able to reproduce a freeze doing following actions:
- Scroll and while scrolling rotate device ([Stacktrace](http://pastebin.com/ivdggiSk))
- Scroll and while scrolling open new Activity

**Comment by erf** [comment]
I've now tested to open an activity whilst scrolling on the latest 532bbd5 but it causes the app to crash.
Example here: https://github.com/erf/mapbox-drawer-glitch/tree/open-activity-whilst-scrolling

**Comment by erf** [comment]
I thought i could try to stop the scroll/fling event when pushing the button but the map keeps moving even after i set:
<pre>
                mMapView.setScrollEnabled(false);
                mMapView.resetPosition();
                mMapView.setCenterCoordinate(mMapView.getCenterCoordinate(), false);
</pre>
i suspect this is some threading / tileloading issue in native lib.

**Comment by erf** [comment]
That is actually another issue, setting map position should cancel scroll

**Comment by erf** [comment]
Are http tile requests canceled when calling onPause on MapView ?

**Comment by erf** [comment]
In the pull request #2297  a call to setCenterCoordinate, calls cancelTransitions, which cancel scrolling before opening an Activity, this seems to stop the app from crashing

**Comment by erf** [comment]
i have still experienced crashes when switching between activities. i'll test some more and create an example if this happens in other scenarios.

**Comment by ljbade** [comment]
@erf The underlying crash I suspect is https://github.com/mapbox/mapbox-gl-native/issues/2295

**Comment by ljbade** [comment]
Interesting yeah, the whole app/activity freezes with anything that either destroys an activity or results in `onPause` while scrolling.

For example if I fling/scroll the map then either rotate the screen or hit the app switch button on the Samsung S6 the app will freeze.

**Comment by ljbade** [comment]
I think that this and https://github.com/mapbox/mapbox-gl-native/issues/1941 are related.

**Comment by bleege** [comment]
To help facilitate debugging and testing (both now and as things go forward), I'm going to create a second map activity in the TestApp.  This will give a common codebase for everyone to experiment with.

**Comment by bleege** [comment]
I added the 2nd Activity to the source code and it behaved just as expected navigating back and forth between the 2.

When I rotated the device, that's when things froze up.  Here's the stack trace:

```
09-15 14:05:56.222    6620-6620/com.mapbox.mapboxgl.testapp I/mbgl﹕ {Main}[OpenGL]: GL Extensions: GL_EXT_debug_marker GL_OES_EGL_image GL_OES_EGL_image_external GL_OES_EGL_sync GL_OES_vertex_half_float GL_OES_framebuffer_object GL_OES_rgb8_rgba8 GL_OES_compressed_ETC1_RGB8_texture GL_AMD_compressed_ATC_texture GL_KHR_texture_compression_astc_ldr GL_OES_texture_npot GL_EXT_texture_filter_anisotropic GL_EXT_texture_format_BGRA8888 GL_OES_texture_3D GL_EXT_color_buffer_float GL_EXT_color_buffer_half_float GL_QCOM_alpha_test GL_OES_depth24 GL_OES_packed_depth_stencil GL_OES_depth_texture GL_OES_depth_textur
09-15 14:05:56.222    6620-6620/com.mapbox.mapboxgl.testapp I/mbgl﹕ {Main}[OpenGL]: GL Extensions: e_cube_map GL_EXT_sRGB GL_OES_texture_float GL_OES_texture_float_linear GL_OES_texture_half_float GL_OES_texture_half_float_linear GL_EXT_texture_type_2_10_10_10_REV GL_EXT_texture_sRGB_decode GL_OES_element_index_uint GL_EXT_copy_image GL_EXT_geometry_shader GL_EXT_tessellation_shader GL_OES_texture_stencil8 GL_EXT_shader_io_blocks GL_OES_shader_image_atomic GL_OES_sample_variables GL_EXT_texture_border_clamp GL_EXT_multisampled_render_to_texture GL_OES_shader_multisample_interpolation GL_EXT_texture_cube_
09-15 14:05:56.222    6620-6620/com.mapbox.mapboxgl.testapp I/mbgl﹕ {Main}[OpenGL]: GL Extensions: map_array GL_EXT_draw_buffers_indexed GL_EXT_gpu_shader5 GL_EXT_robustness GL_EXT_texture_buffer GL_OES_texture_storage_multisample_2d_array GL_OES_sample_shading GL_OES_get_program_binary GL_EXT_debug_label GL_KHR_blend_equation_advanced GL_KHR_blend_equation_advanced_coherent GL_QCOM_tiled_rendering GL_ANDROID_extension_pack_es31a GL_EXT_primitive_bounding_box GL_OES_standard_derivatives GL_OES_vertex_array_object GL_EXT_disjoint_timer_query GL_KHR_debug GL_EXT_sRGB_write_control
09-15 14:05:56.422    6620-7240/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:05:56.532    6620-7240/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:06:03.932    6620-7240/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:08.122    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:08.182    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:08.202    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:08.272    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:08.642    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:08.652    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:08.722    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:08.782    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:09.192    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:09.202    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:09.262    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:09.342    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:09.872    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:09.902    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:09.962    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:10.042    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:10.892    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:10.982    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:11.022    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:11.022    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:11.702    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:11.722    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:11.782    6620-6642/com.mapbox.mapboxgl.testapp I/art﹕ Background sticky concurrent mark sweep GC freed 8426(1425KB) AllocSpace objects, 10(1073KB) LOS objects, 35% free, 4MB/6MB, paused 12.593ms total 39.826ms
09-15 14:11:11.782    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:11.832    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:11.982    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:11.992    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:12.022    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:12.052    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:12.482    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:12.482    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:12.562    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:12.612    6620-6673/com.mapbox.mapboxgl.testapp E/mbgl﹕ {SQLite Cache}[Database](5): database is locked
09-15 14:11:13.552    6620-8782/com.mapbox.mapboxgl.testapp I/mbgl﹕ {Map}[Android]: Not activating as we are not ready
09-15 14:11:13.552    6620-8782/com.mapbox.mapboxgl.testapp I/mbgl﹕ {Map}[Android]: Not deactivating as we are not ready
```

**Comment by bleege** [comment]
I've finished creating `SecondMapActivity` who's whole purpose is life is to exist as a second Activity that can be launched from `MainActivity` and put two `MapView` running Activities in the Android stack concurrently.  The TestApp works exactly as expected without rotating it.  In other words it can roundtrip launching the `SecondMapActivity` and go back to `MainActivity` using the Toolbar many times without any issue.

Please use this for future testing as it'll provide a common way to diagnose and resolve the issue.

![device-2015-09-15-155736](https://cloud.githubusercontent.com/assets/989928/9888483/57ab2fc2-5bc3-11e5-995d-8390b6af5276.png)
_Launch `SecondMapActivity` From Drawer`

![device-2015-09-15-155753](https://cloud.githubusercontent.com/assets/989928/9888492/654b19da-5bc3-11e5-8698-6e7c23dad330.png)
_The Majestic `SecondMapActivity` In All It's Glory_

**Comment by ljbade** [comment]
@bleege nice work, thanks!

**Comment by ljbade** [comment]
I noticed that on Android we create a cache per MapView, where on iOS the same cache is shared with all instances.

Going to try setting that up on Android.

**Comment by ljbade** [comment]
After discussion with @incanus the plan here is to implement common iOS/Android implementation of `MGLFileCache` (a `SQLiteCache` singleton). Using C++'s `shared_ptr` as documented in http://btorpey.github.io/blog/2014/02/12/shared-singletons/

The constructor will take a `std::string` for the platform dependent path to the cache file.

Basically this will fix a problem where there is multiple instances of `SQLiteCache` sharing the same database file which results in locking. Instead a single `SQLiteCache` object is shared among the map instances so there is no file locking.

It seems this problem was fixed at some time on iOS with `MGLFileCache` but this was not implemented on Android.

**Comment by ljbade** [comment]
@bleege Can you give https://github.com/mapbox/mapbox-gl-native/tree/2018-shared-sqlite a test drive?

Testing on the Samsung S6 I don't get the database lock errors. But I still get the freeze on rotation and the occasional crash.

**Comment by erf** [comment]
Exited for this progress! *Cheering* =)

**Comment by kkaefer** [comment]
@ljbade can we use a uv_once/pthread_once?

**Comment by kkaefer** [comment]
Actually never mind, the existing code looks good.

**Comment by zugaldia** [comment]
I'm adding a `SecondMapActivityTest` class to the Espresso test suite to help troubleshoot this.

**Comment by bleege** [comment]
I'm looking into the `2018-shared-sqlite` branch, but ran into an issue with `make android` now from https://github.com/mapbox/mapbox-gl-native/pull/2288.  I'm going to undo it in this branch to test it out.

**Comment by bleege** [comment]
Ok, I've done my time in C++ land and can test this out now.

**Comment by bleege** [comment]
\o/  The database lock errors are now gone.  I was able to rotate the test device from portrait to landscape and back again several times without any problems.!  Great job @ljbade!

Will discuss the NPM issue with @lucaswoj before merging in though.  Hopefully later today.

**Comment by incanus** [comment]
To round this out, we should probably get rid of iOS's `MGLFileCache` and move to this lower-level C++ implementation. I can look at that.

**Comment by incanus** [comment]
To be clear: it's not a blocker and we can ticket off the iOS refactor separately.

**Comment by bleege** [comment]
The plan is to remove 3764b051e4c9944ddbc3bcc9c8fd1beb8a28451e from `2018-shared-sqlite` and then merge @ljbade's fix into `master`.  We can then cherrypick that commit into `release-android-0.1.0` so that both `master` and `release-android-0.1.0` have it and a SNAPSHOT can be rolled because `release-android-0.1.0` does not have the `npm` stuff.

Concurrently @lucaswoj @mikemorris will work on getting `npm` working again so that regular Android development can continue on `master`.

/cc @jfirebaugh

**Comment by incanus** [comment]
Refs https://github.com/mapbox/mapbox-gl-native/pull/2344 re: `npm` problems introduced.

**Comment by bleege** [comment]
Merged.  Now the cherrypick to `release-android-0.1.0`.

**Comment by incanus** [comment]
Speculatively fixed in https://github.com/mapbox/mapbox-gl-native/pull/2345.

**Comment by erf** [comment]
Awsome work !! I would like to test this out but i get the following build error: 
<pre>
deps/run_gyp geojsonvt.gyp -Iconfig.gypi --depth=. -Goutput_dir=. --generator-output=./build -f make
make -C build install
  CXX(target) Release/obj.target/geojsonvt/src/geojsonvt.o
  CXX(target) Release/obj.target/geojsonvt/src/geojsonvt_clip.o
  CXX(target) Release/obj.target/geojsonvt/src/geojsonvt_convert.o
  CXX(target) Release/obj.target/geojsonvt/src/geojsonvt_simplify.o
  CXX(target) Release/obj.target/geojsonvt/src/geojsonvt_tile.o
  LIBTOOL-STATIC Release/libgeojsonvt.a
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: file: Release/obj.target/geojsonvt/src/geojsonvt.o is not an object file (not allowed in a library)
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: file: Release/obj.target/geojsonvt/src/geojsonvt_clip.o is not an object file (not allowed in a library)
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: file: Release/obj.target/geojsonvt/src/geojsonvt_convert.o is not an object file (not allowed in a library)
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: file: Release/obj.target/geojsonvt/src/geojsonvt_simplify.o is not an object file (not allowed in a library)
error: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/libtool: file: Release/obj.target/geojsonvt/src/geojsonvt_tile.o is not an object file (not allowed in a library)
make[3]: *** [Release/libgeojsonvt.a] Error 1
make[2]: *** [install] Error 2
make[1]: *** [config/android-arm-v7.gypi] Error 2
make: *** [android-lib] Error 2
</pre>

**Comment by incanus** [comment]
@erf We are tracking that in https://github.com/mapbox/mapbox-gl-native/issues/2348. For now you can test out the `release-android-v0.1.0` branch, where we cherry-picked the fix, to essentially be based off of an older version of `master` that builds fine.

**Comment by erf** [comment]
Ok, thanks!

**Comment by erf** [comment]
i checked out the 'release-android-v0.1.0' branch, built it and tried it with my test app here: https://github.com/erf/mapbox-drawer-glitch/tree/open-activity-whilst-scrolling
It seems to work much better now, however i experienced a crash after going back and forth between the activites a few times, in the log i only got these:
<pre>
09-17 12:21:23.961  32460-32671/? A/libc﹕ Fatal signal 11 (SIGSEGV), code 1, fault addr 0x3c in tid 32671 (sm-v1/2/2/1.pbf)
09-17 12:28:04.231    3204-3483/? A/libc﹕ Fatal signal 11 (SIGSEGV), code 1, fault addr 0x3c in tid 3483 (-v1/6/33/16.pbf)
</pre>
